File: solutions/python/day5.py
with open("./input.txt", "r") as txt:
  raw: str = txt.read().rstrip()
  txt.close()

rawData = [item.split('\n') for item in raw.split('\n\n')]
rules, pages = [rule.split('|') for rule in rawData[0]], [page.split(',') for page in rawData[1]]

def pageIsValid(page: list[str]) -> bool:
    for rule in rules:
        if rule[0] in page and rule[1] in page:
            # print(rule, page.index(rule[0]), page.index(rule[1]))
            if page.index(rule[0]) > page.index(rule[1]):
                return False
    return True

def fixPage(page: list[str]) -> list[str]:
    for i in range(len(rules) - 1):
        for j in range(i, len(rules)):
            if rules[j][0] in page and rules[j][1] in page:
                if page.index(rules[j][0]) > page.index(rules[j][1]):
                    page[page.index(rules[j][0])], page[page.index(rules[j][1])] = page[page.index(rules[j][1])], page[page.index(rules[j][0])]
    return page

validPages, fixedPages = [], []
for page in pages:
    if pageIsValid(page):
        validPages.append(page)
    else:
        fixedPages.append(fixPage(page))

resultsPart1 = 0
for page in validPages:
    resultsPart1 += int(page[len(page) // 2])
print("Part 1:", resultsPart1)

resultsPart2 = 0
for page in fixedPages:
    resultsPart2 += int(page[len(page) // 2])
print("Part 2:", resultsPart2)

File: solutions/day5.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const [rules, pages] = raw
  .split("\n\n")
  .map((data) =>
    data.split("\n").map((item) => item.replaceAll(",", "|").split("|"))
  );

// Part 1
function isValidPage(page: string[]): boolean {
  for (const rule of rules) {
    if (
      page.includes(rule[1]) &&
      page.indexOf(rule[0]) > page.indexOf(rule[1])
    ) {
      return false;
    }
  }
  return true;
}

// Part 2
function getFixedPage(page: string[]): string[] {
  const fixedPage = [...page];
  for (let i = 0; i < rules.length - 0; i++) {
    for (const rule of rules.slice(i)) {
      if (fixedPage.includes(rule[0]) && fixedPage.includes(rule[1])) {
        const pageIndex0 = fixedPage.indexOf(rule[0]);
        const pageIndex1 = fixedPage.indexOf(rule[1]);
        if (pageIndex0 > pageIndex1) {
          const temp = fixedPage[pageIndex0];
          fixedPage[pageIndex0] = fixedPage[pageIndex1];
          fixedPage[pageIndex1] = temp;
        }
      }
    }
  }
  return fixedPage;
}

const { validPages, fixedPages } = pages.reduce<{
  validPages: string[][];
  fixedPages: string[][];
}>(
  (acc, page) => {
    if (isValidPage(page)) {
      acc.validPages.push(page);
    } else {
      acc.fixedPages.push(getFixedPage(page));
    }
    return acc;
  },
  { validPages: [], fixedPages: [] }
);

const resultsPart1 = validPages.reduce<number>(
  (acc, page) => acc + parseInt(page[Math.floor(page.length / 2)]),
  0
);
console.log("Part 1:", resultsPart1);
const resultsPart2 = fixedPages.reduce<number>(
  (acc, page) => acc + parseInt(page[Math.floor(page.length / 2)]),
  0
);
console.log("Part 2:", resultsPart2);
